# Data Cleaning project
The aim of this project is to clean the data and analyze the included used car listings. 

In this guided project, I'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.


In [1]:
import numpy as np
import pandas as pd

In [2]:
cars=pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
cars.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.
- Text format in columns price, odemetr instead of numeric format.

In [4]:
cars.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
new_columns=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_N', 'postal_code',
       'last_seen']

In [6]:
cars.columns=new_columns

In [7]:
cars.describe(include='all') # to see all text and numeric data

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  pictures_N   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

In [8]:
cars['price']=cars['price'].str.replace('$','').str.replace(',','').astype(int)

In [9]:
cars['odometer']=cars['odometer'].str.replace('km','').str.replace(',','').astype(int)
cars.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

In [10]:
cars.sort_values('price',ascending=False)

date_crawled                                               name  \
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
...                    ...                                                ...   
43510  2016-03-20 21:50:00                                Verkauf_oder_Tausch   
33620  2016-03-26 18:39:00  VW_Jetta_Oldtimer_Youngtimer_Sammler_Retro_1.8...   
39231  2016-03-29 08:57:50                             Suchen_7_Sitzer_//_Bus   
48060  2016-04-02 11:54:30                            Bmw_Verkauf_oder_Tausch   
38354  2016-03-23 04:36:59               Verkaufe_auto_bekommt_kein_tuev_mehr   

       seller offer_type     price  ab_test vehicle_type  registration_year  \
39705  privat    Angebot  99999999  control    limousine               1999   
42221  privat    Angebot  27322222  control    limousine               2014   
27371  privat    Angebot  12345678  control          NaN               2017   
39377  privat    Angebot  12345678  control          NaN               2018   
47598  privat    Angebot  12345678  control    limousine               2001   
...       ...        ...       ...      ...          ...                ...   
43510  privat    Angebot         0     test   kleinwagen               1998   
33620  privat    Angebot         0     test          NaN               2018   
39231  privat    Angebot         0  control          bus               2005   
48060  privat    Angebot         0     test        kombi               1999   
38354  privat    Angebot         0     test          NaN               2000   

         gearbox  power_PS     model  odometer_km  registration_month  \
39705  automatik       224  s_klasse       150000                   9   
42221    manuell       163        c4        40000                   2   
27371        NaN        95     punto       150000                   0   
39377    manuell        95       v40       150000                   6   
47598    manuell       101    vectra       150000                   3   
...          ...       ...       ...          ...                 ...   
43510    manuell       102  a_klasse       150000                   1   
33620  automatik        90       NaN       150000                   3   
39231        NaN         0       NaN        80000                   0   
48060    manuell       150       5er       150000                   3   
38354        NaN         0       NaN       150000                   0   

      fuel_type           brand unrepaired_damage           ad_created  \
39705    benzin   mercedes_benz               NaN  2016-03-22 00:00:00   
42221    diesel         citroen               NaN  2016-03-08 00:00:00   
27371       NaN            fiat               NaN  2016-03-09 00:00:00   
39377       NaN           volvo              nein  2016-03-08 00:00:00   
47598    benzin            opel              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
43510    benzin   mercedes_benz              nein  2016-03-20 00:00:00   
33620       NaN      volkswagen              nein  2016-03-26 00:00:00   
39231       NaN  sonstige_autos              nein  2016-03-29 00:00:00   
48060    benzin             bmw              nein  2016-04-02 00:00:00   
38354       NaN           mazda               NaN  2016-03-23 00:00:00   

       pictures_N  postal_code            last_seen  
39705           0        73525  2016-04-06 05:15:30  
42221           0        76532  2016-03-08 20:39:05  
27371           0        96110  2016-03-09 15:45:47  
39377           0        14542  2016-04-06 23:17:31  
47598           0         4356  2016-03-31 18:56:54  
..

result looks like we have some mistakes:
- max price
- price 123456789
- price 0 ( it's mean that cars was free)

So, let's remove this rows

In [11]:
cars_clean = cars[(cars['price']>1000) & (cars['price']<11111111)]

In [12]:
cars_clean.describe()

price  registration_year      power_PS    odometer_km  \
count  3.799400e+04        37994.00000  37994.000000   37994.000000   
mean   7.870498e+03         2005.76591    129.599858  122491.051219   
std    5.728079e+04           87.39647    216.422360   40899.095541   
min    1.039000e+03         1000.00000      0.000000    5000.000000   
25%    2.300000e+03         2001.00000     80.000000  100000.000000   
50%    4.499000e+03         2005.00000    116.000000  150000.000000   
75%    8.999000e+03         2009.00000    160.000000  150000.000000   
max    1.000000e+07         9999.00000  17700.000000  150000.000000   

       registration_month  pictures_N   postal_code  
count        37994.000000     37994.0  37994.000000  
mean             6.016318         0.0  51745.480023  
std              3.584050         0.0  25679.509038  
min              0.000000         0.0   1067.000000  
25%              3.000000         0.0  31246.000000  
50%              6.000000         0.0  50938.000000  
75%              9.000000         0.0  72555.000000  
max             12.000000         0.0  99998.000000

In [13]:
cars_clean[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

First 10 characters is data

In [14]:
cars_clean['date_crawled']=cars_clean['date_crawled'].str[:10]

<ipython-input-14-bd5eb5edb7e0>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cars_clean['date_crawled']=cars_clean['date_crawled'].str[:10]


In [16]:
cars_clean['ad_created']=cars_clean['ad_created'].str[:10]
cars_clean['last_seen']=cars_clean['last_seen'].str[:10]

<ipython-input-16-af3664af7921>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cars_clean['ad_created']=cars_clean['ad_created'].str[:10]
<ipython-input-16-af3664af7921>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cars_clean['last_seen']=cars_clean['last_seen'].str[:10]


In [17]:
cars_clean['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025715
2016-03-06    0.014002
2016-03-07    0.035453
2016-03-08    0.032347
2016-03-09    0.032479
2016-03-10    0.033268
2016-03-11    0.032953
2016-03-12    0.037638
2016-03-13    0.016055
2016-03-14    0.036769
2016-03-15    0.033505
2016-03-16    0.029031
2016-03-17    0.030163
2016-03-18    0.012923
2016-03-19    0.035111
2016-03-20    0.038243
2016-03-21    0.037322
2016-03-22    0.032584
2016-03-23    0.032058
2016-03-24    0.028794
2016-03-25    0.030294
2016-03-26    0.033137
2016-03-27    0.031452
2016-03-28    0.035348
2016-03-29    0.034084
2016-03-30    0.032953
2016-03-31    0.031400
2016-04-01    0.034558
2016-04-02    0.036374
2016-04-03    0.039243
2016-04-04    0.036769
2016-04-05    0.013239
2016-04-06    0.003237
2016-04-07    0.001500
Name: date_crawled, dtype: float64

In [56]:
cars_clean['registration_year'].describe()

count    37994.00000
mean      2005.76591
std         87.39647
min       1000.00000
25%       2001.00000
50%       2005.00000
75%       2009.00000
max       9999.00000
Name: registration_year, dtype: float64

we see that max year 9999 and min 1000 - is mistakes

let's make df with registration years between 1920-2022

In [18]:
cars_clean['registration_year'].value_counts()

2005    2770
2006    2647
2004    2585
2003    2447
2007    2257
        ... 
1938       1
1939       1
5911       1
4500       1
1952       1
Name: registration_year, Length: 91, dtype: int64

In [19]:
cars_cleaned=cars_clean[cars_clean['registration_year'].between(1920,2022)]

In [20]:
cars_cleaned.describe()

price  registration_year      power_PS    odometer_km  \
count  3.798100e+04       37981.000000  37981.000000   37981.000000   
mean   7.869523e+03        2004.519944    129.624918  122530.475764   
std    5.729021e+04           7.322932    216.444823   40850.504490   
min    1.039000e+03        1927.000000      0.000000    5000.000000   
25%    2.300000e+03        2001.000000     81.000000  100000.000000   
50%    4.499000e+03        2005.000000    116.000000  150000.000000   
75%    8.999000e+03        2009.000000    160.000000  150000.000000   
max    1.000000e+07        2019.000000  17700.000000  150000.000000   

       registration_month  pictures_N   postal_code  
count        37981.000000     37981.0  37981.000000  
mean             6.017772         0.0  51744.442800  
std              3.583426         0.0  25681.154667  
min              0.000000         0.0   1067.000000  
25%              3.000000         0.0  31246.000000  
50%              6.000000         0.0  50937.000000  
75%              9.000000         0.0  72555.000000  
max             12.000000         0.0  99998.000000

In [26]:
brand=cars_cleaned['brand'].value_counts().head(10).index

In [28]:
brands={}
for b in brand:
    brands[b]=cars_cleaned[cars_cleaned['brand']==b]['price'].mean()

In [29]:
brands

{'volkswagen': 6898.993934142114,
 'bmw': 9390.74366554054,
 'mercedes_benz': 9296.579121398205,
 'audi': 10286.987041036717,
 'opel': 4254.891907855877,
 'ford': 5857.823449524672,
 'renault': 3625.2169274537696,
 'peugeot': 4006.038684719536,
 'fiat': 4033.7395061728394,
 'skoda': 6894.382889200561}

In [30]:
bmp_series = pd.Series(brands)

In [31]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
volkswagen      6898.993934
bmw             9390.743666
mercedes_benz   9296.579121
audi           10286.987041
opel            4254.891908
ford            5857.823450
renault         3625.216927
peugeot         4006.038685
fiat            4033.739506
skoda           6894.382889

In [34]:
mileage={}
for b in brand:
    mileage[b]=cars_cleaned[cars_cleaned['brand']==b]['odometer_km'].mean()

In [43]:
mileage_series = pd.Series(mileage)
df['mileage'] = mileage_series

In [44]:
df

mean_price        mileage
volkswagen      6898.993934  125771.849468
bmw             9390.743666  131952.069257
mercedes_benz   9296.579121  130042.512990
audi           10286.987041  127846.922246
opel            4254.891908  123758.121677
ford            5857.823450  119287.007696
renault         3625.216927  121052.631579
peugeot         4006.038685  121900.386847
fiat            4033.739506  108086.419753
skoda           6894.382889  109712.482468

In [49]:
cars_cleaned['date_crawled']

NameError: name 'datetime' is not defined

In [50]:
import datetime as dt

In [58]:
dates=[]
for date in cars_cleaned['date_crawled']:
    dates.append(dt.datetime.strptime(date,"%Y-%m-%d"))

cars_cleaned['date_crawled']=dates

<ipython-input-58-f1fe7d0ffa02>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cars_cleaned['date_crawled']=dates


In [ ]:
car